components/script/script_thread.rs

```diff
@@ -1471,28 +1476,292 @@ impl ScriptThread {
+
+                // TODO: run the update intersection observations steps(#31021).
+
+                // TODO: mark paint timing(https://w3c.github.io/paint-timing).
 
-        for (id, document) in self.documents.borrow().iter() {
-            // Only process a resize if layout is idle.
-            if let Some((size, size_type)) = document.window().steal_resize_event() {
-                resizes.push((id, size, size_type));
+                // TODO: Update the rendering: consolidate all reflow calls into one here?(#31871)
+
+                // TODO: process top layer removals(https://drafts.csswg.org/css-position-4/#process-top-layer-removals)
             }
         }
+    }
 
-        for (id, size, size_type) in resizes {
-            self.handle_event(id, ResizeEvent(size, size_type));
+    /// <https://html.spec.whatwg.org/multipage/#event-loop-processing-model:rendering-opportunity>
+    fn rendering_opportunity(&self, pipeline_id: PipelineId) {
+        *self.last_render_opportunity_time.borrow_mut() = time::precise_time_ns();
+
+        // Note: the pipeline should be a navigable with a rendering opportunity,
+        // and we should use this opportunity to queue one task for each navigable with
+        // an opportunity in this script-thread.
+        let document = match self.documents.borrow().find_document(pipeline_id) {
+            Some(document) => document,
+            None => {
+                return warn!(
+                    "Trying to update the rendering for closed pipeline {}.",
+                    pipeline_id
+                )
+            },
+        };
+        let window = document.window();
+        let task_manager = window.task_manager();
+        let rendering_task_source = task_manager.rendering_task_source();
+        let canceller = task_manager.task_canceller(TaskSourceName::Rendering);
+
+        // Batch rendering ops.
+        if !*self.has_queued_update_the_rendering_task.borrow() {
+            *self.has_queued_update_the_rendering_task.borrow_mut() = true;
+            // Queues a task to update the rendering.
+            // <https://html.spec.whatwg.org/multipage/#event-loop-processing-model:queue-a-global-task>
+            let _ = rendering_task_source.queue_with_canceller(
+                task!(update_the_rendering: move || {
+                    // Note: spec says to queue a task using the navigable's active window,
+                    // but then updates the rendering for all docs in the same event-loop.
+                    SCRIPT_THREAD_ROOT.with(|root| {
+                        if let Some(script_thread) = root.get() {
+                            let script_thread = unsafe {&*script_thread};
+                            script_thread.update_the_rendering();
+                        }
+                    })
+                }),
+                &canceller,
+            );
         }
+    }
+
+    /// Handle incoming control messages.
```

Comment:
Nit: I would say IPC here for a little more clarity.

```suggestion
    /// Handle incoming IPC messages.
```

Reply:
The task queue is not IPC.